### ISSUE TITLE
Make debugging / logging variables in a circuit less error-prone

### ISSUE THREAD

**Issue by mitschabaude** [o1-labs/snarkyjs#489]
feedback by @gretzke 

```
- vielleicht sollte man toString/toJSON ganz entfernen (oder zumindest dem Entwickler nicht zugängig machen) damit man gar nicht in Versuchung kommt etwas zu loggen
- Eine dedicated log funktion anbieten (e.g., snarky.log()) die 
Circuit.asProver(() => {
  console.log(field.toString())
});
```

Paraphrasing:
* remove functions like `field.toString()` / `.toJSON()` because the aren't allowed inside circuits
* instead, expose a simple `debug` / `log` function which abstracts this away
  - example: `field.debug(x => console.log(x))`, where `x` is a string
  - or maybe: `debug(log => { log(field); log(bool) })` which logs variables in a nice format

**Comment by gretzke** [comment]
I think such a wrapper would also work well: 
```js
function log(...args: any[]) {
  Circuit.asProver(() => {
    console.log(...args.map((x) => x.toString()));
  });
}
```
calling it like this:
```js
snarky.log('field', Field.zero, 'bool', Bool(true), 'uint', new UInt64(2));
```
should result in
```
> field 0 bool true uint 2
```
It would allow developers to log values intuitively with the same syntax they are used from `console.log` without breaking the compilation. `toString()` methods should be protected so that the functions cannot be called by the dev directly, because the first intuition of every dev would be logging like this `console.log(myField.toString());`

Another option would be exposing a dedicated log function on the `Field`, and `CircuitValue` classes that also take an optional string for logging, e.g.,:

```js
const myField = Field(5);
myField.log("the value of my field is:");
> the value of my field is: 5
```

**Comment by mitschabaude** [comment]
I like the simplicity of your `snarky.log`, and similarity with `console.log` :D

I wonder whether the innermost function should not be `x => x.toString()`, but `x => JSON.stringify(x)` or `x => JSON.parse(JSON.stringify(x))`, because that will work on all circuit values, and print something nice for plain JS arrays / objects / primitive types as well

**Comment by gretzke** [comment]
The json representation makes sense, I know people also like logging values like this:
```js
console.log({a,b})
> { a: 1, b: 2 }
```
I feel like this would work great with the json parsing

**Comment by bkase** [comment]
`unsafeToStringOutsideCircuit()`

**Comment by gretzke** [comment]
Is there any other reason to convert a provable value to a string except for logging?

**Comment by mitschabaude** [comment]
@gretzke yes, serializing for IO, & in general process it as a normal JS value outside the circuit. It's used all the time, it's just not supposed to be used in circuits

**Comment by mitschabaude** [comment]
@bkase I propose to not change this for now, but instead change the error message to something that explains the matter very well, and to revisit this when we work on separating constants and variables.

The reasoning is the the constant-variable separation could mean we can keep a nice API, so it feels premature to introduce that ugly / dangerous API in the meantime
